map causes a compiler bug error.

### Describe the bug

when trying to declare a  map with a key of type string and a value of type datatypes.Stack<int>{}
 program doesn't compile and state that this is a compiler bug.

```v
import datatypes

p := map[string]datatypes.Stack<int>{}
println(p)
```


### Expected Behavior

I expect a map with the specified key value type to be declared.

### Current Behavior

Throws this error and program is not compiled.
 ```
/tmp/v_1000/t.17442098459935223959.tmp.c:2244: warning: implicit declaration of function 'indent_datatypes__Stack_T_int_str'
/tmp/v_1000/t.17442098459935223959.tmp.c:2244: error: cannot convert 'int' to 'struct string'
```

### Reproduction Steps

Try to declare a map with key of type string and value of type `datatypes.Stack<int>` or `datatypes.Queue<int>` 

### Possible Solution

Throw a friendly error if such operation is not possible like when trying to use value of type [void](https://github.com/vlang/v/blob/7e9e2ff45994c3f6c348e8e6b181fd9ea91df209/vlib/v/parser/parse_type.v#L138)

### Additional Information/Context

_No response_

### V version

V 0.3.2 7e9e2ff

### Environment details (OS name and version, etc.)

```
OS: linux, Pop!_OS 22.04 LTS
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-4210U CPU @ 1.70GHz
CC version: cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0

getwd: /home/mp/AOC/day2
vmodules: /home/mp/.vmodules
vroot: /home/mp/Downloads/v
vexe: /home/mp/Downloads/v/v
vexe mtime: 2022-12-05 06:59:16
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.2 ffc73bf

Git version: git version 2.34.1
Git vroot status: Error: fatal: bad revision 'HEAD'
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -709,8 +709,18 @@ fn (mut g Gen) gen_str_for_map(info ast.Map, styp string, str_fn_name string) {
 		val_sym = g.table.sym(val_typ)
 	}
 	val_styp := g.typ(val_typ)
-	elem_str_fn_name := val_styp.replace('*', '') + '_str'
-	if !val_sym.has_method('str') {
+	mut elem_str_fn_name := val_styp.replace('*', '') + '_str'
+	if val_sym.has_method_with_generic_parent('str') {
+		match mut val_sym.info {
+			ast.Struct, ast.Interface, ast.SumType {
+				if val_sym.info.generic_types.len > 0 {
+					elem_str_fn_name = g.generic_fn_name(val_sym.info.concrete_types,
+						elem_str_fn_name)
+				}
+			}
+			else {}
+		}
+	} else {
 		g.get_str_fn(val_typ)
 	}
 
@@ -742,7 +752,7 @@ fn (mut g Gen) gen_str_for_map(info ast.Map, styp string, str_fn_name string) {
 	} else if val_sym.kind == .string {
 		tmp_str := str_intp_sq('*(${val_styp}*)DenseArray_value(&m.key_values, i)')
 		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, ${tmp_str});')
-	} else if should_use_indent_func(val_sym.kind) && !val_sym.has_method('str') {
+	} else if should_use_indent_func(val_sym.kind) && !val_sym.has_method_with_generic_parent('str') {
 		ptr_str := '*'.repeat(val_typ.nr_muls())
 		g.auto_str_funcs.writeln('\t\tstrings__Builder_write_string(&sb, indent_${elem_str_fn_name}(*${ptr_str}(${val_styp}*)DenseArray_value(&m.key_values, i), indent_count));')
 	} else if val_sym.kind in [.f32, .f64] {

--- a/vlib/v/tests/string_map_with_generic_struct_value_test.v
+++ b/vlib/v/tests/string_map_with_generic_struct_value_test.v
@@ -1,0 +1,7 @@
+import datatypes
+
+fn test_string_map_with_generic_struct_value() {
+	p := map[string]datatypes.Stack[int]{}
+	println(p)
+	assert '${p}' == '{}'
+}

